### PR TITLE
discard and save error after editing a note

### LINKED ISSUE #2738: Save and discard dialog has no affect
Platform: Mobile
OS specifics: No 

## Steps to reproduce

1.  Make a new note or edit a note.
2.  Click back arrow key before it gets saved automatically.
3. You will get a dialog that has options of saving and discarding, but clicking discard doesn't discard changes.

## Describe what you expected to happen

I expected either shouldn't have discard options or it should discard changes after I click on that.

@laurent22  @tessus  can you please tell how should I work on this, I mean, should I remove discard option or discard changes after clicking discard button.
save_dialog

### PR DESCRIPTION
**Pull request by ishantgupta777**
This PR closes #2738. The notes get save automatically, with no need for dialog.
@laurent22  can you look at this PR. I commented on the lines which were showing the dialog box. Do you want me to delete those lines and the functions to show the dialog box?

### CODE REVIEW

**Comment by mic704b** [comment]
Does this PR contain unrelated code changes from another PR?  If so, please fix.

**Comment by ishantgupta777** [comment]
> Does this PR contain unrelated code changes from another PR? If so, please fix.

Oh okay okay, I will remove those lines. I forgot to remove those lines. Thanks.

**Comment by laurent22** [comment]
@ishantgupta777, please do a git diff before committing to see exactly what's in the commit.

Another issue is that you didn't implement spec, even though it was just one sentence: "The dialog should be removed **and the note saved automatically instead**". The note is not saved automatically in your code.

**Comment by ishantgupta777** [comment]
> @ishantgupta777, please do a git diff before committing to see exactly what's in the commit.
> 
> Another issue is that you didn't implement spec, even though it was just one sentence: "The dialog should be removed **and the note saved automatically instead**". The note is not saved automatically in your code.



> @ishantgupta777, please do a git diff before committing to see exactly what's in the commit.
> 
> Another issue is that you didn't implement spec, even though it was just one sentence: "The dialog should be removed **and the note saved automatically instead**". The note is not saved automatically in your code.

I got your point @laurent22 , I am sorry I missed that. I removed the extra lines and fixing my other PRs also. And saving the notes automatically also done. Please have a look. Thanks.

**Comment by laurent22** [comment]
Thanks for the update. As we don't have automated tests for this, please could you list the manual tests you performed to validate this feature and the results?

**Comment by ishantgupta777** [comment]
@PackElend  can you please add gsoc2020 tag in this PR.

**Comment by ishantgupta777** [comment]
> Thanks for the update. As we don't have automated tests for this, please could you list the manual tests you performed to validate this feature and the results?

Sir, I didn't think of any testing scenario in this case. I tested manually with the emulator, everything is fine and notes are getting saved. I tested with already present notes and editing them, by making new notes and by changing different screens. Everything looks good. But can you suggest me some case for which I can write unit testing? I will write it.

**Comment by laurent22** [comment]
We don't have unit tests for this so it would only be manual testing. I can think of several test cases you could do manually:

- Open an existing note, change it, wait a bit and go back. Is the note saved?
- Open an existing note, change it, go back quickly. Is the note saved?
- Create a new note, don't change anything, go back. Is the note saved? (it shouldn't)
- Create a new note, change something, go back quickly. Is the note saved? (it should)
- Create a new note, change something, wait a bit, go back. Is the note saved?

Ideally you would think of these test cases yourself and present them in the pull request.

**Comment by ishantgupta777** [comment]
> We don't have unit tests for this so it would only be manual testing. I can think of several test cases you could do manually:
> 
> * Open an existing note, change it, wait a bit and go back. Is the note saved?
> * Open an existing note, change it, go back quickly. Is the note saved?
> * Create a new note, don't change anything, go back. Is the note saved? (it shouldn't)
> * Create a new note, change something, go back quickly. Is the note saved? (it should)
> * Create a new note, change something, wait a bit, go back. Is the note saved?
> 
> Ideally you would think of these test cases yourself and present them in the pull request.

Thank you so much, sir, this is my first time contributing to this big organization.  I will keep these suggestions in mind and make sure from next time onward I will submit my PR  with proper manual/unit tests and you don't need to get disturbed because of these problems.
dialog_test_gif

**Comment by ishantgupta777** [comment]
dialog_test_gif

**Comment by laurent22** [comment]
Please provide the test results in writing. It's very hard to follow what's going on in these gif.

**Comment by ishantgupta777** [comment]
Results : 
1. After opening an existing note and change something, notes are getting saved instantly and after waiting some time also.
2. After making a new note and changing something also saves the note instantly or after some time also. 
3. If I make a new note and didn't change anything and go back, it makes a new note ( untitled note) which is because of notes getting saved automatically, It's not checking if it's empty or not.
It's just saving them, this is also the case before removing the dialog in this PR.  So either we have to take that case into consideration separately or leave it as it is.
So, all the results are as expected.

**Comment by laurent22** [comment]
@ishantgupta777, I've asked to test 5 cases but you only tested 3. I've just tested the branch and I see you broke something:

When I create a new note, then don't change anything, and click Back, normally it shouldn't save the note (it means the user wanted to create a new note, but decided not to). With your change, it always save the note so it would create a bunch of Untitled notes. You would have found this out had you tested the changes as suggested.

So that we can't merge this I'm afraid.

**Comment by PackElend** [comment]
I labeled it as bug to show that it can be fixed, we can change it to a new label

**Comment by ishantgupta777** [comment]
> @ishantgupta777, I've asked to test 5 cases but you only tested 3. I've just tested the branch and I see you broke something:
> 
> When I create a new note, then don't change anything, and click Back, normally it shouldn't save the note (it means the user wanted to create a new note, but decided not to). With your change, it always saves the note so it would create a bunch of Untitled notes. You would have found this out had you tested the changes as suggested.
> 
> So that we can't merge this I'm afraid.

Sir, I am really sorry, I missed that case, I must have exchanged it with some other case. I have fixed it and this time I checked all the 5 cases properly and all the test cases were succesfull. Please have a look at this now. Thanks
